File: examples/eval-python/eval.py
"""
Example Python evaluation function for Shimmy-WASM (CPython-WASI resident backend).

Contract
--------
evaluation_function(response, answer, params) -> dict
    Called for "eval" requests.
    Must return a dict with at least:
        is_correct  bool
        feedback    str

preview_function(response, answer, params) -> dict  [optional]
    Called for "preview" requests.
    If not defined, the runner falls back to evaluation_function.
    Should return a preview/hint without revealing the full solution.

Shimmy's configured lifecycle must prevent global-variable mutations from
leaking between requests. The invocation counter below is runtime-lane reset
evidence, not application state.
"""


_guest_invocation_count = 0


def evaluation_function(response, answer, params=None):
    """Numeric equality check with configurable absolute tolerance."""
    global _guest_invocation_count
    _guest_invocation_count += 1
    params = params or {}
    tolerance = float(params.get("tolerance", 1e-9))

    try:
        r = float(response)
        a = float(answer)
    except (TypeError, ValueError) as e:
        return {
            "is_correct": False,
            "feedback": f"Error: Could not parse values as numbers: {e}",
            "guest_invocation_count": _guest_invocation_count,
        }

    abs_err = abs(r - a)
    is_correct = abs_err <= tolerance

    if is_correct:
        return {
            "is_correct": True,
            "feedback": f"Correct! {r} matches {a} within tolerance {tolerance}.",
            "absolute_error": abs_err,
            "guest_invocation_count": _guest_invocation_count,
        }
    else:
        return {
            "is_correct": False,
            "feedback": (
                f"Incorrect. Got {r}, expected {a}. "
                f"Absolute error: {abs_err:.6e} (tolerance: {tolerance:.6e})."
            ),
            "absolute_error": abs_err,
            "guest_invocation_count": _guest_invocation_count,
        }


def preview_function(response, answer, params=None):
    """Return a hint without revealing whether the answer is correct."""
    params = params or {}
    tolerance = float(params.get("tolerance", 1e-9))

    try:
        r = float(response)
    except (TypeError, ValueError):
        return {
            "preview": f"Could not parse response '{response}' as a number.",
        }

    return {
        "preview": (
            f"Your answer is {r}. "
            f"The checker uses absolute tolerance {tolerance:.2e}."
        ),
    }

File: examples/eval-js/eval.js
/**
 * Example JavaScript evaluation function for Shimmy-WASM (QuickJS/javy backend).
 *
 * This file contains the user-defined evaluation logic.  The runner (runner.js)
 * embeds this source at build time and executes it inside a fresh Function()
 * scope for every request, preventing global-variable mutations from leaking
 * between requests.
 *
 * Contract
 * --------
 *   evaluationFunction(response, answer, params) → object
 *     Called for "eval" requests.
 *     Must return an object with at least:
 *       is_correct  boolean
 *       feedback    string
 *
 *   previewFunction(response, answer, params) → object  [optional]
 *     Called for "preview" requests.
 *     If not defined, runner falls back to evaluationFunction.
 *     Should return a hint without revealing whether the answer is correct.
 *
 * Supported params keys
 * ---------------------
 *   tolerance   float    Absolute tolerance for numeric comparison (default 1e-6)
 */

"use strict";

function evaluationFunction(response, answer, params) {
  params = params || {};

  var tolerance = parseFloat(params.tolerance !== undefined ? params.tolerance : 1e-6);

  var r = parseFloat(response);
  var a = parseFloat(answer);

  if (isNaN(r) || isNaN(a)) {
    return {
      is_correct: false,
      feedback: "Could not parse values as numbers (response=" + response + ", answer=" + answer + ").",
    };
  }

  var absErr = Math.abs(r - a);
  var isCorrect = absErr <= tolerance;

  if (isCorrect) {
    return {
      is_correct: true,
      feedback: "Correct! " + r + " matches " + a + ".",
      absolute_error: absErr,
    };
  } else {
    return {
      is_correct: false,
      feedback:
        "Incorrect. Got " +
        r +
        ", expected " +
        a +
        ". Absolute error: " +
        absErr.toExponential(6) +
        " (tolerance: " +
        tolerance.toExponential(6) +
        ").",
      absolute_error: absErr,
    };
  }
}

function previewFunction(response, answer, params) {
  params = params || {};
  var tolerance = parseFloat(params.tolerance !== undefined ? params.tolerance : 1e-6);
  var r = parseFloat(response);

  if (isNaN(r)) {
    return {
      preview: "Could not parse response '" + response + "' as a number.",
    };
  }

  return {
    preview:
      "Your answer is " +
      r +
      ". The checker uses absolute tolerance " +
      tolerance.toExponential(2) +
      ".",
  };
}
